pkg/operator/ceph/csi/csidriver.go

```diff
@@ -0,0 +1,116 @@
+*/
+
+package csi
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	v1k8scsi "k8s.io/api/storage/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/storage/v1"
+)
+
+type csiDriver interface {
+	createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, name, fsGroupPolicy string) error
+	reCreateCSIDriverInfo(ctx context.Context) error
+	deleteCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, name string) error
+}
+
+type v1CsiDriver struct {
+	csiDriver *v1k8scsi.CSIDriver
+	csiClient v1.CSIDriverInterface
+}
+
+// createCSIDriverInfo Registers CSI driver by creating a CSIDriver object
+func (d v1CsiDriver) createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, name, fsGroupPolicy string) error {
+	attach := true
+	mountInfo := false
+	// Create CSIDriver object
+	csiDriver := &v1k8scsi.CSIDriver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: v1k8scsi.CSIDriverSpec{
+			AttachRequired: &attach,
+			PodInfoOnMount: &mountInfo,
+		},
+	}
+	if fsGroupPolicy != "" {
+		policy := v1k8scsi.FSGroupPolicy(fsGroupPolicy)
+		csiDriver.Spec.FSGroupPolicy = &policy
+	}
+	csidrivers := clientset.StorageV1().CSIDrivers()
+	driver, err := csidrivers.Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			_, err = csidrivers.Create(ctx, csiDriver, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+			logger.Infof("CSIDriver object created for driver %q", name)
+		}
+		return err
+	}
+
+	// As FSGroupPolicy field is immutable, should be set only during create time.
+	// if the request is to change the FSGroupPolicy, we are deleting the CSIDriver object and creating it.
+	if driver.Spec.FSGroupPolicy != nil && csiDriver.Spec.FSGroupPolicy != nil && *driver.Spec.FSGroupPolicy != *csiDriver.Spec.FSGroupPolicy {
+		return d.reCreateCSIDriverInfo(ctx)
+	}
+
+	// For csidriver we need to provide the resourceVersion when updating the object.
+	// From the docs (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata)
+	// > "This value MUST be treated as opaque by clients and passed unmodified back to the server"
+	csiDriver.ObjectMeta.ResourceVersion = driver.ObjectMeta.ResourceVersion
+	_, err = csidrivers.Update(ctx, csiDriver, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	logger.Infof("CSIDriver object updated for driver %q", name)
+	return nil
+}
+
+func (d v1CsiDriver) reCreateCSIDriverInfo(ctx context.Context) error {
+	csiDriver := d.csiDriver
+	csiClient := d.csiClient
+	err := csiClient.Delete(ctx, csiDriver.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete CSIDriver object for driver %q", csiDriver.Name)
+	}
+	logger.Infof("CSIDriver object deleted for driver %q", csiDriver)
```

Comment:
sorry I missed this one,Log only driver name?

Reply:
I was wondering for returning errors it's fine to use just the driver name but for logging it is better if we log the whole driver. What do you think 🤔?

Reply:
In other recreate function it's logging only driver name. If the whole object is logged it might flood the operator logs. To keep simple and consistent I will go for just name logging. @travisn any suggestion?

Reply:
Since it is just an Info level log, I'm fine with logging just the name. Will wait for a suggestion from @travisn

Reply:
Yes sounds good just to log the driver name

Reply:
Updated now.